src/zope/i18nmessageid/_zope_i18nmessageid_message.c

```diff
@@ -174,7 +174,10 @@ Message_new(PyTypeObject* type, PyObject* args, PyObject* kwds)
     } else if (mapping != NULL) {
         /* Ensure that our mapping is immutable */
         new_msg->mapping = PyDictProxy_New(mapping);
-    } else {}
+    } else if (other != NULL && new_msg->mapping == other->mapping) {
+        /* borrowed? */
+        Py_XINCREF(new_msg->mapping);
```

Comment:
@tseaver I'm able to reproduce the segfault by running the manylinux image locally on my aarch64 machine. gdb shows the error on this line.

```
Program received signal SIGSEGV, Segmentation fault.
Message_new (type=0x1daf1a0, args=<optimized out>, kwds=<optimized out>) at src/zope/i18nmessageid/_zope_i18nmessageid_message.c:179
179	        Py_XINCREF(new_msg->mapping);
```

Reply:
@davisagli Thanks for tracking that down!  Now I just need to figure out how that could possibly segfault.

`Py_XINCREF` is defined so in Python 3.8, 3.9, and 3.10:
```python
static inline void _Py_XINCREF(PyObject *op)
{
    if (op != NULL) {
        Py_INCREF(op);
    }
}

#define Py_XINCREF(op) _Py_XINCREF(_PyObject_CAST(op))
```

`Py_XINCREF` is defined so in Python 3.11 / 3.12:
```python
static inline void Py_XINCREF(PyObject *op)
{
    if (op != _Py_NULL) {
        Py_INCREF(op);
    }
}
#if !defined(Py_LIMITED_API) || Py_LIMITED_API+0 < 0x030b0000
#  define Py_XINCREF(op) Py_XINCREF(_PyObject_CAST(op))
#endif
```

I cannot see how that could segfault, unless the "borrowed" mapping is already corrupt.

Reply:
@tseaver Maybe a refcounting bug elsewhere which means the borrowed mapping was already gc'd?

Reply:
@davisagli 

Even weirder:  the [failed job](https://github.com/zopefoundation/zope.i18nmessageid/actions/runs/10633940916/job/29480138539?pr=58) is blowing up in the doctests for `docs/narr.rst`.

It pretty much *has* to be [this line](https://github.com/zopefoundation/zope.i18nmessageid/blob/287330aeca6527f01e72b874607d4f0a2c59c461/docs/narr.rst?plain=1#L118):
```python
  >>> new_robot = Message(robot, mapping={'name': 'Bender'})
```
as that is the only example which copies one message from another, but the offending line is not reachable here, because `mapping` is passed in explicitly (and `robot.mapping` is `None`, anyway).

Reply:
@davisagli  I've got some hope that d198112 fixes that segfault.

Reply:
@tseaver It fixed it on my machine.